City-of-Helsinki/hel-fi-drupal-grants
# feat: AU-2017: Implemented application draft validation

# [AU-2017](https://helsinkisolutionoffice.atlassian.net/browse/AU-2017)

## What was done
This pull request implements application draft validation. For now, the validation is only implemented for the following forms:
- 29 (registered community)
- 48 (all roles)
- 51 (registered community)
- 53 (registered community)
- 56 (registered community)
- 63 (registered community)
- 64 (registered community)

The validation is done by visiting a forms "View" page, (e.g. "/hakemus/LOCALT-048-0033996/katso"), and verifying that the submitted form data is present on said page.

## Implemented changes
- Some basic profile data is now derived from `profile_input_data.ts`. This data is used when setting up user profiles for private users, registered communities and unregistered communities. The same data is also used in the application_data_X files. This change was implemented so that we can verify that the user data correctly ends up on an applications "View" page after an application has either been saved as a draft or submitted.

- The `FormFieldWithRemove` interface has been extended with three (3) new keys: `viewPageSelector`, `viewPageFormatter` and `viewPageSkipValidation`.

  ![Screenshot 2024-01-18 at 15 34 46](https://github.com/City-of-Helsinki/hel-fi-drupal-grants/assets/117807015/4d942800-4120-4940-997f-76aa51312aef)


  - `viewPageSelector`: Allows a `FormField` to define a class by which its content can be located on an applications "View" page. Note that this only needs to be used if a class can't be constructed from the items key itself (more information further down).
  
  - `viewPageFormatter`: Allows a `FormField` to define a callback function to format its data before it is compared against the results on the "View" page (more on these callback functions further down).
  
  - `viewPageSkipValidation`: Allows a `FormField` to define if the field should be skipped when performing validation.
  
     ![Screenshot 2024-01-18 at 16 13 34](https://github.com/City-of-Helsinki/hel-fi-drupal-grants/assets/117807015/8c74b3e2-ef3c-4ddc-a37d-18d68d95d9cb)

 
- The file `view_page_formatters.ts` contains callback functions that can be set to the previously mentioned `viewPageFormatter` key. These callback functions format a form fields input data so that it match the output on an applications "View" page. For example, an input date of "2023-11-01" can be formatted to "01.11.2023" with the help of the `viewPageFormatDate` function.

  ![Screenshot 2024-01-18 at 16 16 03](https://github.com/City-of-Helsinki/hel-fi-drupal-grants/assets/117807015/72853e64-aa76-4892-834c-460c45094413)


- The file `view_page_helpers.ts` contains helper functions for the "View" page. For now, it only contains the `viewPageBuildSelectorForItem` functions, which attempts to formulate a class by which to target a specific form field on an applications "View" page. Note that class constructed by this function will always be used to target a form item as long as the form item has not specified its own class with the `viewPageSelector` key.

  ![Screenshot 2024-01-18 at 16 18 47](https://github.com/City-of-Helsinki/hel-fi-drupal-grants/assets/117807015/06fc9605-a738-4d96-8bcb-f4ede7a63836)

- A new "type" has been added to the `fillInputField` function: `data-drupal-selector-sequential`. The new type is similar to the `data-drupal-selector`, the only difference being that it fills a form field with the help of the `pressSequentially()` method instead of the `fill()` method. This change was needed in order to get formatted number fields on our forms to work correctly. When using the normal `fill()` method, entering the number "1023" into, for example, a "budget" field, will produce the output "1".

  ![Screenshot 2024-01-18 at 16 19 24](https://github.com/City-of-Helsinki/hel-fi-drupal-grants/assets/117807015/2d139747-214b-4dc9-a260-ee6a0aec208e)

- The validation itself has been implemented inside the `validateDraft()` function in the `validation_helpers.ts` file. If you want a more detailed output of the validation, you can remove the comments in the `logDraftValidationResults()` function.

- Many of the tests had hard-coded data in them. These have been changed to use the input data instead.

## Things I've noticed
- When performing validation, I would some times run into issues where two strings that seemed identical did not match according to Playwright. The problem turned out to be different kinds of space characters in the input text and the output text (one used ASCII Code 32 for the space and the other ASCII Code 160). If you run into similar issues, this might be worth looking into.

- This seems to happen randomly, but some times attachments fail to upload when the tests are being run, which obviously causes the validation to fail. Just know that you are not the first one experiencing this problem if it ever happens to you.

## How to install
Make sure your instance is up and running on the correct branch.
  * `git checkout feature/AU-2017-verify-form-submission-e2e`
  * `make fresh`
  * `make drush-cr`

## How to test
- Edit `fetchLatestProfileByType()` and make it return `false`. This will force new profile creation, and the usage of the data in `PROFILE_INPUT_DATA`.

Now you can run the tests:
- `make test-pw-p PROJECT=forms-48`.
- `make test-pw registered_community_29 `.
- `make test-pw registered_community_51 `.
- `make test-pw registered_community_53 `.
- `make test-pw registered_community_56 `.
- `make test-pw registered_community_63 `.
- `make test-pw registered_community_64 `.

NOTE!: The validation of registered_community_64 is probably going to fail, since there is a bug on the form (correct information not showing on the draft page for registered communities). This has been reported on the 26.1.2024.

The output should look similar to this when validation succeeds:

![Screenshot 2024-01-18 at 16 23 38](https://github.com/City-of-Helsinki/hel-fi-drupal-grants/assets/117807015/4ebd0841-2c68-4cbb-bd75-eb3f33970324)

The output should look similar to this when validation fails:

![Screenshot 2024-01-26 at 13 52 34](https://github.com/City-of-Helsinki/hel-fi-drupal-grants/assets/117807015/2c107dc6-baf1-4c92-a148-4e8362bdf2cb)


[AU-2017]: https://helsinkisolutionoffice.atlassian.net/browse/AU-2017?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
